### ISSUE TITLE
stb_image and stb_image_write: allow to use zlib for png compression

### ISSUE THREAD

**Issue by guillaumechereau** [nothings/stb#113]
I noticed that the zlib compression in stb_image_write was rather poor compared to what the actual zlib library would do.  For example saving an rgba image filled with the gray color (128,128,128,255) prodices a png of size 42KB.  When I use zlib to compress the data instead of stb algorithm, I get 6.4KB.

Many projects already use zlib, so I think having the option to use it would be nice, if that is not too much work.

**Comment by nothings** [comment]
This is documented. Use libpng?

**Comment by nothings** [comment]
Although, note to self: worth investigating why that doesn't compress smaller.

**Comment by guillaumechereau** [comment]
libpng works, but I still prefer stb_image_write interface, plus the fact that it is a single file.
Isn't the compression problem due to the fact that you use a fixed huffman tree?

**Comment by nothings** [comment]
If you hook it up to zlib it's no longer a single file.

I guess my stance is that if you want something to ship to customers, you might as well use libpng; stb_image_write is intended for quick-and-dirty stuff where you can always run pngcrush etc if you really care.

The fixed tree definitely limits the effective compression, but a basically-constant file should probably compress better anyway, so I wonder if the match processing isn't right.

**Comment by r-lyeh** [comment]
@guillaumechereau, lodepng is somewhere between stb_image_write and libpng. you might give it a run as well.

**Comment by jarikomppa** [comment]
miniz appears to be a small, unlicensed zlib implementation - https://code.google.com/p/miniz/ - could possibly be used as an alternative?

**Comment by guillaumechereau** [comment]
On the platforms I am mostly interested in (android), zlib is always present, so it is basically free to use it.  I thought adding a test macro like STB_HAS_ZLIB to the code could have been a solution.  But I agree with @nothings that if the stb implementation has a problem, it would be better to fix it first.

**Comment by jarikomppa** [comment]
I meant, if a better zlib implementation is wanted in stb_image_write and the desire is to keep it in one file, that might be one shortcut..

**Comment by DanielGibson** [comment]
miniz actually contains a png compression function that is easy to use, see https://code.google.com/p/miniz/source/browse/trunk/miniz.c#822  
Maybe just use that instead of stb_image_write?

(Of course it would be great if stb_image_write's png compression was closer to libpng, but then the optimization should be done in the lib itself and not by using other dependencies - especially if the dependency can compress png images itself anyway)

**Comment by nothings** [comment]
It looks to me like the miniz doesn't use PNG filters, so it's just compressing the raw image data as is, which isn't going to be very good compression either, but along a different axis. But maybe I'm misreading the code.

**Comment by DanielGibson** [comment]
I thought I read claims that miniz's png compression is at least as good as libpng, but I can't find them anymore so I may misremember.

Maybe someone could do a quick comparison between miniz, stb_image_write and maybe libpng, but I'm feeling lazy :-/

**Comment by DanielGibson** [comment]
Ok, felt a bit less lazy today: http://wp.me/pEPJ4-5U compares `stbi_write_png()` with miniz's png compressor, LodePNG's `lodepng_encode_file()` and those 100 functions libpng needs to get the job done.

TL;DR: You were right, miniz's PNG compressor is not really better than yours: Sometimes it's a bit better, sometimes worse, and both compress a lot less than libpng and LodePNG.

So if anyone wants to create PNG images with better compression I'd suggest either improving stbi_write_png() or using LodePNG (which is almost as easy to integrate and use as stb_image_write).
(For the sake of completeness: LodePNG can also load PNGs, but is slower at it than stb_image)

**Comment by DanielGibson** [comment]
I hacked `stbi_write_png_to_mem()` to use miniz's `mz_compress2()` instead of `stbi_zlib_compress()` - the compression ratio gets very close to libpng then.

The corresponding lines with a bit of context, if anyone wants to try it:
```
   STBIW_FREE(line_buffer);
   
   // zlib = stbi_zlib_compress(filt, y*(x*n+1), &zlen, 8); // increase 8 to get smaller but use more memory
   mz_ulong buflen = mz_compressBound(y*(x*n+1));
   zlib = STBIW_MALLOC(buflen);
   if(!zlib || mz_compress2(zlib, &buflen, filt, y*(x*n+1), 9))
   {
      STBIW_FREE(filt);
      STBIW_FREE(zlib);
      return 0;
   }
   
   zlen = buflen;
   
   STBIW_FREE(filt);
   if (!zlib) return 0;
```
(+ you need to #include "miniz.c" somewhere in the implementation part)

So I guess the png compression could indeed be improved a lot by borrowing miniz code (which is also public domain).

(From the miniz side it would be easier actually, they'd just have to borrow the `stbi_write_png_to_mem()` code + 15lines of helpers, which is much shorter than the miniz compression code)

**Comment by nothings** [comment]
@guillaumechereau do you remember what dimensions your gray test image was?

**Comment by godratio** [comment]
This is would be absolutely valuable for stb_image to upgrade here.  Its just this one thing away from being useable production side.

**Comment by DanielGibson** [comment]
I've implemented this, see #470

**Comment by nothings** [comment]
fixed by #470

**Comment by 1vanK** [comment]
From my tests stb not only saves pngs in a larger size but also much slower than miniz

**Comment by 1vanK** [comment]
6 images 1024x1024, release build, visual studio:

Miniz MZ_BEST_SPEED

![изображение](https://github.com/user-attachments/assets/245020bb-6e88-4888-931a-96e4166a4bf3)

```
[2025-01-10 20:24:59] INFO: SpriteFont::SpriteFont(const FontSettingsSimple&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu/Ubuntu-R.ttf | Generated im 55 ms
[2025-01-10 20:24:59] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_0.png | Saved im 11 ms
[2025-01-10 20:24:59] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_1.png | Saved im 11 ms
[2025-01-10 20:24:59] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_2.png | Saved im 11 ms
[2025-01-10 20:24:59] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_3.png | Saved im 12 ms
[2025-01-10 20:24:59] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_4.png | Saved im 11 ms
[2025-01-10 20:24:59] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_5.png | Saved im 8 ms
```

Miniz MZ_DEFAULT_LEVEL:

![изображение](https://github.com/user-attachments/assets/f6c81fba-a9cf-4c0f-8b25-642aba4bef47)

```
[2025-01-10 20:24:20] INFO: SpriteFont::SpriteFont(const FontSettingsSimple&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu/Ubuntu-R.ttf | Generated im 53 ms
[2025-01-10 20:24:20] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_0.png | Saved im 31 ms
[2025-01-10 20:24:20] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_1.png | Saved im 32 ms
[2025-01-10 20:24:20] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_2.png | Saved im 33 ms
[2025-01-10 20:24:20] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_3.png | Saved im 34 ms
[2025-01-10 20:24:20] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_4.png | Saved im 32 ms
[2025-01-10 20:24:20] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_5.png | Saved im 16 ms
```

Miniz MZ_BEST_COMPRESSION

![изображение](https://github.com/user-attachments/assets/56738af7-9351-4abb-a4c6-ced063b23c82)

```
[2025-01-10 20:23:20] INFO: SpriteFont::SpriteFont(const FontSettingsSimple&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu/Ubuntu-R.ttf | Generated im 53 ms
[2025-01-10 20:23:20] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_0.png | Saved im 92 ms
[2025-01-10 20:23:21] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_1.png | Saved im 93 ms
[2025-01-10 20:23:21] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_2.png | Saved im 95 ms
[2025-01-10 20:23:21] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_3.png | Saved im 99 ms
[2025-01-10 20:23:21] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_4.png | Saved im 93 ms
[2025-01-10 20:23:21] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_5.png | Saved im 39 ms
```

Stb:

![изображение](https://github.com/user-attachments/assets/bb55e1e8-8c76-46f6-a8b4-195c4cd0ac56)


```
[2025-01-10 20:22:05] INFO: SpriteFont::SpriteFont(const FontSettingsSimple&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu/Ubuntu-R.ttf | Generated im 59 ms
[2025-01-10 20:22:05] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_0.png | Saved im 46 ms
[2025-01-10 20:22:05] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_1.png | Saved im 46 ms
[2025-01-10 20:22:05] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_2.png | Saved im 47 ms
[2025-01-10 20:22:05] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_3.png | Saved im 50 ms
[2025-01-10 20:22:05] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_4.png | Saved im 46 ms
[2025-01-10 20:22:05] INFO: Image::save_png(const StrUtf8&) | F:/my_games/dviglo 2d/build_vs/result/engine_test_data/fonts/ubuntu-r_20_simple_5.png | Saved im 55 ms
```
